crates/matrix-sdk/src/event_cache/mod.rs

```diff
@@ -505,51 +483,100 @@ impl RoomEventCacheInner {
+        }
 
-            // Clear internal state (events, pagination tokens, etc.).
-            store.clear_room(self.room.room_id()).await?;
+        Ok(())
+    }
 
-            // Propagate to observers.
-            let _ = self.sender.send(RoomEventCacheUpdate::Clear);
-        }
+    async fn handle_left_room_update(&self, updates: LeftRoomUpdate) -> Result<()> {
+        self.handle_timeline(updates.timeline, Vec::new(), Vec::new(), updates.ambiguity_changes)
+            .await?;
+        Ok(())
+    }
+
+    // Remove existing events, and append a set of events to the room cache and
+    // storage, notifying observers.
+    async fn replace_all_events_by(
+        &self,
+        events: Vec<SyncTimelineEvent>,
+        prev_batch: Option<String>,
+        account_data: Vec<Raw<AnyRoomAccountDataEvent>>,
+        ephemeral: Vec<Raw<AnySyncEphemeralRoomEvent>>,
+        ambiguity_changes: BTreeMap<OwnedEventId, AmbiguityChange>,
+    ) -> Result<()> {
+        // Acquire the lock.
+        let mut room_events = self.events.write().await;
+
+        // Reset the events.
+        room_events.reset();
 
-        // Add all the events to the backend.
-        trace!("adding new events");
-        self.append_events(
-            store,
-            timeline.events,
-            timeline.prev_batch,
+        // Propagate to observers.
+        let _ = self.sender.send(RoomEventCacheUpdate::Clear);
+
+        // Push the new events.
+        self.append_events_locked_impl(
+            Some(room_events),
+            events,
+            prev_batch,
             account_data,
             ephemeral,
             ambiguity_changes,
         )
-        .await?;
-
-        Ok(())
+        .await
     }
 
-    async fn handle_left_room_update(
+    /// Append a set of events to the room cache and storage, notifying
+    /// observers.
+    async fn append_new_events(
         &self,
-        store: &dyn EventCacheStore,
-        updates: LeftRoomUpdate,
+        events: Vec<SyncTimelineEvent>,
+        prev_batch: Option<String>,
+        account_data: Vec<Raw<AnyRoomAccountDataEvent>>,
+        ephemeral: Vec<Raw<AnySyncEphemeralRoomEvent>>,
+        ambiguity_changes: BTreeMap<OwnedEventId, AmbiguityChange>,
     ) -> Result<()> {
-        self.handle_timeline(
-            store,
-            updates.timeline,
-            Vec::new(),
-            Vec::new(),
-            updates.ambiguity_changes,
+        // Push the new events.
+        self.append_events_locked_impl(
+            None,
+            events,
+            prev_batch,
+            account_data,
+            ephemeral,
+            ambiguity_changes,
         )
-        .await?;
-        Ok(())
+        .await
     }
 
-    /// Append a set of events to the room cache and storage, notifying
-    /// observers.
-    async fn append_events(
+    /// Append a set of events, with an attached lock.
+    ///
+    /// If the lock `room_events` is `None`, one will be created.
```

Comment:
Yeah, it was my first design. But then I had doubt because if `events` is empty, `prev_batch` is empty`, `ephemeral`, `account_data` and `ambiguity_changes` are all empty, we simply return early. I don't know if it's a good idea to take the lock for no reason in this particular case.

We can also assume that this case is pretty rare as far as I know. I'm modifying the code according to your feedback, let's see how it goes.